sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images/ZE_experimental_bindless_images.asciidoc

```diff
@@ -0,0 +1,621 @@
+    uint32_t numWaitEvents,                             /// [in][optional] number of events to wait on before launching
+    ze_event_handle_t *phWaitEvents                     /// [in][optional][range(0, numWaitEvents)] handle of the events to wait on before launching
+)
+
+/// @brief Appends a command to copy image memory from the device to the host
+///
+/// @details
+///     - This functions allows for the copying of data from image memory before
+///       an image handle is created
+///
+/// @returns
+///     - ::ZE_RESULT_SUCCESS
+///     - ::ZE_RESULT_ERROR_UNINITIALIZED
+///     - ::ZE_RESULT_ERROR_DEVICE_LOST
+///     - ::ZE_RESULT_ERROR_OUT_OF_HOST_MEMORY
+///     - ::ZE_RESULT_ERROR_OUT_OF_DEVICE_MEMORY
+///     - ::ZE_RESULT_ERROR_INVALID_NULL_HANDLE
+///         + nullptr == hCommandList
+///         + nullptr == hDstImageMem
+///     - ::ZE_RESULT_ERROR_INVALID_NULL_POINTER
+///         + nullptr == srcptr
+///     - ::ZE_RESULT_ERROR_INVALID_SYNCHRONIZATION_OBJECT
+///     - ::ZE_RESULT_ERROR_INVALID_SIZE
+///         + (nullptr == phWaitEvents) && (0 < numWaitEvents)
+ZE_APIEXPORT ze_result_t ZE_APICALL
+zeCommandListAppendImageMemoryCopyToHostExp(
+    ze_command_list_handle_t hCommandList,              /// [in] handle of command list
+    void *dstptr,                                       /// [in] pointer to destination memory to copy to
+    const ze_image_mem_handle_t hSrcImageMem,           /// [in] handle of source image memory to copy from
+    const ze_image_region_t *pSrcRegion,                /// [in][optional] source region descriptor
+    ze_event_handle_t hSignalEvent,                     /// [in][optional] handle of the event to signal on completion
+    uint32_t numWaitEvents,                             /// [in][optional] number of events to wait on before launching
+    ze_event_handle_t *phWaitEvents                     /// [in][optional][range(0, numWaitEvents)] handle of the events to wait on before launching
+)
+
+/// @brief Appends a command to copy image USM memory from the host to the device
+///
+/// @details
+///     - This functions allows for the copying of data to image memory before
+///       an image handle is created
+///
+/// @returns
+///     - ::ZE_RESULT_SUCCESS
+///     - ::ZE_RESULT_ERROR_UNINITIALIZED
+///     - ::ZE_RESULT_ERROR_DEVICE_LOST
+///     - ::ZE_RESULT_ERROR_OUT_OF_HOST_MEMORY
+///     - ::ZE_RESULT_ERROR_OUT_OF_DEVICE_MEMORY
+///     - ::ZE_RESULT_ERROR_INVALID_NULL_HANDLE
+///         + nullptr == hCommandList
+///         + nullptr == hDstImageMem
+///     - ::ZE_RESULT_ERROR_INVALID_NULL_POINTER
+///         + nullptr == srcptr
+///     - ::ZE_RESULT_ERROR_INVALID_SYNCHRONIZATION_OBJECT
+///     - ::ZE_RESULT_ERROR_INVALID_SIZE
+///         + (nullptr == phWaitEvents) && (0 < numWaitEvents)
+ZE_APIEXPORT ze_result_t ZE_APICALL
+zeCommandListAppendImageUSMCopyFromHostExp(
```

Comment:
we dont need this and neither CopyToHostExp. This is a usm pointer, so we can the zeCommandListAppendMemoryCopy

Reply:
thanks @isaacault . Here we are passing pitched memory, which is USM, right? L0 internally should be able to detect whether the pointer is regular USM memory or USM pitched memory, and do what it needs. ze_image_region_t is the same as ze_copy_region_t, and for that we have zeCommandListAppendMemoryCopyRegion. Unless we need to pass some extra info to the copy, then we could use that one already. All that I see below is ze_image_region_t, and zeCommandListAppendMemoryCopyRegion would cover that.

Reply:
@isaacault :

> We'll address this in the next revision

was this next revision already posted? if not, when do you think we will have it?

Reply:
thanks @isaacault .

I will move this now to the L0 spec repo.